### PR TITLE
Ecards

### PR DESCRIPTION
**Pull request by PavelBulat**

### CODE REVIEW

File: .docker/docker-compose.override.default.yml [hunk 1]
```diff
@@ -27,6 +27,12 @@ services:
       - ../:/var/www/html:cached
 
   codecept:
+## Codeception container with xDebug.
+#    image: pavelbulat/xdebug_codeception
```
Comment (ikorgik) [inline .docker/docker-compose.override.default.yml]: Currently it conflicts with Kate's changes, can you please check with her?

File: composer.json [hunk 2]
```diff
@@ -95,6 +95,9 @@
       },
       "drupal/focal_point": {
         "Support Media Library": "https://www.drupal.org/files/issues/2018-09-21/focal_point-integrate_with_media_library-3001262-4.patch"
+      },
+      "drupal/commerce_decoupled_checkout": {
+        "Added new hook_alter": "https://www.drupal.org/files/issues/2019-02-22/commerce_decoupled_checkout-added_hook_alter-3033373-5.patch"
```
Comment (ikorgik) [inline composer.json:100]: Do you know when this patch can be committed?

File: modules/custom/falcon_gift_ecards/falcon_gift_ecards.module [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      try {
+        // Creating ecard entity.
+        $entity = $storage->create($entityData);
+        $entity->save();
+
+        // Preparing post data to create an order.
+        $data['order']['order_items'][$key]['field_card'] = $entity->id();
+      }
+      catch (Exception $e) {
+        Drupal::logger('falcon_gift_ecards')->alert($e->getMessage());
```
Comment (ikorgik) [inline modules/custom/falcon_gift_ecards/falcon_gift_ecards.module]: I would define more meaningful alert message and attached $e->getMessage() additionally. 
e.g: `->alert("Can't create an event entity. Error: " . $e->getMessage());`

File: modules/custom/falcon_gift_ecards/src/EventSubscriber/FalconOrderCompletedSubscriber.php [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+          $field_cards = $item->get('field_card')->getValue();
+
+          // Send ecards for completed order.
+          foreach ($field_cards as $field_card) {
+            $card = $this->storage->load($field_card['target_id']);
+            if (!empty($card) &&
+              !empty($card->get('field_status')->getValue()) &&
+              !$card->get('field_status')->getValue()[0]['value']) {
+
+              $this->mailSender->send($card);
```
Comment (ikorgik) [inline modules/custom/falcon_gift_ecards/src/EventSubscriber/FalconOrderCompletedSubscriber.php]: I would recommend to wrap this part with try-catch

File: modules/custom/falcon_gift_ecards/falcon_gift_ecards.module [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    case 'ecard_mail':
+      if (isset($params['headers'])) {
+        $message['headers'] = array_merge($message['headers'], $params['headers']);
+      }
+
+      $message['format'] = 'text/html';
+      $message['from'] = $params['from'];
+      $message['subject'] = $params['subject'];
+      $message['body'][] = $params['body'];
+  }
```
Comment (ikorgik) [inline modules/custom/falcon_gift_ecards/falcon_gift_ecards.module:61]: Do we need 'break' at the end of case?

File: modules/custom/falcon_gift_ecards/src/MailSender.php [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+   * @return bool
+   *   TRUE if success.
+   */
+  public function send($ecardItem) {
+
+    if ($ecardItem->bundle() != 'ecard') {
+      return FALSE;
+    }
+
+    $to = $ecardItem->get('field_recipient_email')->value;
```
Comment (ikorgik) [inline modules/custom/falcon_gift_ecards/src/MailSender.php]: I would recommend to wrap this function with try-catch

File: modules/custom/falcon_gift_ecards/src/MailSenderInterface.php [hunk 7]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace Drupal\falcon_gift_ecards;
+
+/**
+ * Interface MailSenderInterface.
+ *
+ * @package Drupal\falcon_gift_ecards
+ */
+interface MailSenderInterface {
```
Comment (ikorgik) [inline modules/custom/falcon_gift_ecards/src/MailSenderInterface.php]: Do we need an empty interface file here at all?

**Review by ikorgik — CHANGES_REQUESTED** [review]

File: composer.json (hunk 2 above)
Comment (PavelBulat) [reply composer.json:100]: @spleshka When you know when this patch can be committed? =))

File: .docker/docker-compose.override.default.yml (hunk 1 above)
Comment (PavelBulat) [reply .docker/docker-compose.override.default.yml]: Merged.

File: modules/custom/falcon_gift_ecards/falcon_gift_ecards.module (hunk 3 above)
Comment (PavelBulat) [reply modules/custom/falcon_gift_ecards/falcon_gift_ecards.module]: The message added.

File: modules/custom/falcon_gift_ecards/src/EventSubscriber/FalconOrderCompletedSubscriber.php (hunk 4 above)
Comment (PavelBulat) [reply modules/custom/falcon_gift_ecards/src/EventSubscriber/FalconOrderCompletedSubscriber.php]: Done.

File: modules/custom/falcon_gift_ecards/falcon_gift_ecards.module (hunk 5 above)
Comment (PavelBulat) [reply modules/custom/falcon_gift_ecards/falcon_gift_ecards.module:61]: The 'break'  added.

File: modules/custom/falcon_gift_ecards/src/MailSender.php (hunk 6 above)
Comment (PavelBulat) [reply modules/custom/falcon_gift_ecards/src/MailSender.php]: Done.

File: modules/custom/falcon_gift_ecards/src/MailSenderInterface.php (hunk 7 above)
Comment (PavelBulat) [reply modules/custom/falcon_gift_ecards/src/MailSenderInterface.php]: Deleted.

**Review by ikorgik — APPROVED** [review]

**Pull request merged**